mlflow/sklearn/utils.py

```diff
@@ -103,6 +119,253 @@ def _get_args_for_score(score_func, fit_func, fit_args, fit_kwargs):
+
+    For (4) accuracy score:
+    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html
+    we choose the parameter `normalize` to be `True` to output the percentage of accuracy,
+    as opposed to `False` that outputs the absolute correct number of sample prediction
+
+    We log additional metrics if certain classifier has method `predict_proba`
+    (5) log loss:
+    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.log_loss.html#sklearn.metrics.log_loss
+    (6) roc_auc_score:
+    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#r4bb7c4558997-5
+    By default, for roc_auc_score, we pick `average` to be `weighted`, `multi_class` to be `ovo`,
+    to make the output more insensitive to dataset imbalance.
+
+    Steps:
+    1. Extract X and y_true from fit_args and fit_kwargs, and compute y_pred.
+    2. If the sample_weight argument exists in fit_func (accuracy_score by default
+    has sample_weight), extract it from fit_args or fit_kwargs as
+    (y_true, y_pred, ...... sample_weight), otherwise as (y_true, y_pred, ......)
+    3. return a dictionary of metric(name, value)
+
+    :param fitted_estimator: The already fitted classifier
+    :param fit_args: Positional arguments given to fit_func.
+    :param fit_kwargs: Keyword arguments given to fit_func.
+    :return: dictionary of (function name, computed value)
+    """
+    import sklearn
+
+    fit_arg_names = _get_arg_names(fitted_estimator.fit)
+    X, y_true, y_pred = _get_samples_labels_and_predictions(
+        fitted_estimator, fit_args, fit_kwargs, fit_arg_names
+    )
+    sample_weight = (
+        _get_sample_weight(fit_arg_names, fit_args, fit_kwargs)
+        if _SAMPLE_WEIGHT in fit_arg_names
+        else None
+    )
+
+    classifier_metrics = [
+        _SklearnMetric(
+            name="training_precision_score",
+            function=sklearn.metrics.precision_score,
+            arguments=dict(
+                y_true=y_true, y_pred=y_pred, average="weighted", sample_weight=sample_weight
+            ),
+        ),
+        _SklearnMetric(
+            name="training_recall_score",
+            function=sklearn.metrics.recall_score,
+            arguments=dict(
+                y_true=y_true, y_pred=y_pred, average="weighted", sample_weight=sample_weight
+            ),
+        ),
+        _SklearnMetric(
+            name="training_f1_score",
+            function=sklearn.metrics.f1_score,
+            arguments=dict(
+                y_true=y_true, y_pred=y_pred, average="weighted", sample_weight=sample_weight
+            ),
+        ),
+        _SklearnMetric(
+            name="training_accuracy_score",
+            function=sklearn.metrics.accuracy_score,
+            arguments=dict(
+                y_true=y_true, y_pred=y_pred, normalize=True, sample_weight=sample_weight
+            ),
+        ),
+    ]
```

Comment:
Thse lines can be simplifed to:

```python
metric_funcs = {
    sklearn.metrics.precision_score: {"average": "weighted"},
    sklearn.metrics.accuracy_score: {"average": "weighted"},
    # ...
}

classifier_metrics = [
    _SklearnMetric(
        name="training_{}".format(func.__name__),
        function=func,
        arguments=dict(y_true=y_true, y_pred=y_pred, sample_weight=sample_weight, **kwargs),
    )
    for func, kwargs in metric_funcs.items()
]
```

Reply:
I have actually though about simplify these lines, but I kinda feel this optimization is a bit overkill and impair the readability. As there are some special cases that y_pred=y_pred_proba.
If we only simplify the above common metrics, it seems a bit inconsistent though.
Listing like now looks pretty straightforward, and if in the future we want to add other metrics we may not need think about refactor the code (As you suggestion suggest that all functions always take in y_true=y_true, y_pred=y_pred, sample_weight=sample_weight,)

Reply:
Makes sense :)

Reply:
Makes sense :)